homeassistant/components/qbus/coordinator.py

```diff
@@ -0,0 +1,274 @@
+            )
+
+            await mqtt.async_publish(self.hass, request.topic, request.payload)
+
+        if len(controller.outputs) > 0:
+            async_call_later(self.hass, self._WAIT_TIME, request_state)
+
+    def _request_controller_state(self, controller: QbusMqttDevice) -> None:
+        async def request_controller_state(_: datetime) -> None:
+            _LOGGER.debug("%s - Requesting controller state", self._entry.unique_id)
+            request = self._message_factory.create_device_state_request(controller)
+            await mqtt.async_publish(self.hass, request.topic, request.payload)
+
+        async_call_later(self.hass, self._WAIT_TIME, request_controller_state)
+
+
+class QbusConfigCoordinator:
+    """Class responsible for Qbus config updates."""
+
+    _qbus_config: QbusDiscovery | None = None
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize config coordinator."""
+
+        self._hass = hass
+        self._message_factory = QbusMqttMessageFactory()
+        self._topic_factory = QbusMqttTopicFactory()
+        self._cleanup_callbacks: list[CALLBACK_TYPE] = []
+
+        self._cleanup_callbacks.append(
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.shutdown)
+        )
+
+    @classmethod
+    def get_or_create(cls, hass: HomeAssistant) -> QbusConfigCoordinator:
+        """Get the coordinator and create if necessary."""
+        if (coordinator := hass.data.get(QBUS_KEY)) is None:
+            coordinator = cls(hass)
+            hass.data[QBUS_KEY] = coordinator
+
+        return coordinator
+
+    def shutdown(self, event: Event | None = None) -> None:
+        """Shutdown Qbus config coordinator."""
+        _LOGGER.debug("Shutting down Qbus config coordinator")
+        while self._cleanup_callbacks:
+            cleanup_callback = self._cleanup_callbacks.pop()
+            cleanup_callback()
+
+    async def async_subscribe_to_config(self) -> None:
+        """Subscribe to config changes."""
+        config_topic = self._topic_factory.get_config_topic()
+        _LOGGER.debug("Subscribing to %s", config_topic)
+
+        self._cleanup_callbacks.append(
+            await mqtt.async_subscribe(self._hass, config_topic, self._config_received)
+        )
+
+    async def async_get_or_request_config(self) -> QbusDiscovery | None:
+        """Get or request Qbus config."""
+        _LOGGER.debug("Requesting Qbus config")
+
+        # Config already available
+        if self._qbus_config:
+            _LOGGER.debug("Qbus config already available")
+            return self._qbus_config
+
+        if not await async_wait_for_mqtt_client(self._hass):
+            _LOGGER.debug("MQTT client not ready yet")
+            return None
+
+        # Request config
+        _LOGGER.debug("Publishing config request")
+        await mqtt.async_publish(
+            self._hass, self._topic_factory.get_get_config_topic(), b""
+        )
+
+        return self._qbus_config
```

Comment:
at this point `_qbus_config` is None, right? should we `return None` directly to make that explicit?

Reply:
There is a chance it is not None, when the subscription fires between the first line of this method and this statement. Qbus could also publish config updates on its own.